src/event_generator.hpp

```diff
@@ -0,0 +1,209 @@
+    // Return the next event
+    // Should return the same event if called multiple times without calling
+    // event_generator::pop().
+    virtual postsynaptic_spike_event next() = 0;
+    virtual void pop() = 0;
+    virtual void reset() = 0;
+    virtual ~event_generator() {};
+    virtual void advance(time_type t) = 0;
+};
+
+inline
+postsynaptic_spike_event terminal_pse() {
+    return postsynaptic_spike_event{cell_member_type{0,0}, max_time, 0};
+}
+
+// Generator that feeds events that are specified with a vector
+struct vector_backed_generator: public event_generator {
+    using pse = postsynaptic_spike_event;
+    vector_backed_generator(pse_vector events):
+        events_(std::move(events)),
+        it_(events_.begin())
+    {
+        if (!std::is_sorted(events_.begin(), events_.end())) {
+            util::sort(events_);
+        }
+    }
+
+    postsynaptic_spike_event next() override {
+        return it_==events_.end()? terminal_pse(): *it_;
+    }
+
+    void pop() override {
+        if (it_!=events_.end()) {
+            ++it_;
+        }
+    }
+
+    void reset() override {
+        it_ = events_.begin();
+    }
+
+    void advance(time_type t) override {
+        it_ = std::lower_bound(events_.begin(), events_.end(), t, event_time_less());
+    }
+
+private:
+    std::vector<postsynaptic_spike_event> events_;
+    std::vector<postsynaptic_spike_event>::const_iterator it_;
+};
+
+// Generator that for events in a generic sequence
+template <typename Seq>
+struct seq_generator: public event_generator {
+    using pse = postsynaptic_spike_event;
+    seq_generator(Seq& events):
+        events_(events),
+        it_(std::begin(events_))
+    {
+        EXPECTS(std::is_sorted(events_.begin(), events_.end()));
+    }
+
+    postsynaptic_spike_event next() override {
+        return it_==events_.end()? terminal_pse(): *it_;
+    }
+
+    void pop() override {
+        if (it_!=events_.end()) {
+            ++it_;
+        }
+    }
+
+    void reset() override {
+        it_ = events_.begin();
+    }
+
+    void advance(time_type t) override {
+        it_ = std::lower_bound(events_.begin(), events_.end(), t, event_time_less());
+    }
+
+private:
+
+    const Seq& events_;
+    typename Seq::const_iterator it_;
+};
+
+// Generates a stream of events
+//  * with the first event at time t_start
+//  * subsequent events at t_start+n*dt, ∀ n ∈ [0, ∞)
+//  * with a set target and weight
+struct regular_generator: public event_generator {
+    using pse = postsynaptic_spike_event;
+
+    regular_generator(time_type tstart, time_type dt, cell_member_type target, float weight):
+        t_start_(tstart), step_(0), dt_(dt),
+        target_(target), weight_(weight)
+    {}
+
+    postsynaptic_spike_event next() override {
+        return {target_, t_start_+(step_*dt_), weight_};
+    }
+
+    void pop() override {
+        ++step_;
+    }
+
+    void advance(time_type t0) override {
+        t0 = t0<t_start_? t_start_: t0;
```

Comment:
I would put a short comment how this works.

Reply:
Ha, you are right! I should just use `std::max(to, t_start)`